bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/api/MyWarmupApi.java

```diff
@@ -0,0 +1,202 @@
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.openhab.binding.warmup.internal.WarmupBindingConstants;
+import org.openhab.binding.warmup.internal.handler.MyWarmupConfigurationDTO;
+import org.openhab.binding.warmup.internal.model.auth.AuthRequestDTO;
+import org.openhab.binding.warmup.internal.model.auth.AuthResponseDTO;
+import org.openhab.binding.warmup.internal.model.query.QueryResponseDTO;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+/**
+ * The {@link MyWarmupApi} class contains code specific to calling the My Warmup API.
+ *
+ * @author James Melville - Initial contribution
+ */
+@NonNullByDefault
+public class MyWarmupApi {
+
+    private static final Gson GSON = new Gson();
+
+    private final Logger logger = LoggerFactory.getLogger(MyWarmupApi.class);
+    private final HttpClient httpClient;
+
+    private MyWarmupConfigurationDTO configuration;
+    private @Nullable String authToken;
+    private int failCount = 0;
+
+    /**
+     * Construct the API client
+     *
+     * @param httpClient HttpClient to make HTTP Calls
+     * @param configuration Thing configuration which contains API credentials
+     */
+    public MyWarmupApi(final HttpClient httpClient, MyWarmupConfigurationDTO configuration) {
+        this.httpClient = httpClient;
+        this.configuration = configuration;
+    }
+
+    /**
+     * Update the configuration, trigger a refresh of the access token
+     *
+     * @param configuration contains username and password
+     */
+    public void setConfiguration(MyWarmupConfigurationDTO configuration) {
+        authToken = null;
+        this.configuration = configuration;
+    }
+
+    private Boolean validateSession() throws MyWarmupApiException {
+        if (authToken == null) {
+            return authenticate();
+        }
+        return true;
+    }
+
+    private Boolean authenticate() throws MyWarmupApiException {
+        try {
+            String body = GSON.toJson(new AuthRequestDTO(configuration.username, configuration.password,
+                    WarmupBindingConstants.AUTH_METHOD, WarmupBindingConstants.AUTH_APP_ID));
+
+            ContentResponse response = callWarmup(WarmupBindingConstants.APP_ENDPOINT, body, false);
+
+            AuthResponseDTO ar = GSON.fromJson(response.getContentAsString(), AuthResponseDTO.class);
+
+            if (ar.getStatus().getResult().equals("success")) {
+                authToken = ar.getResponse().getToken();
+                failCount = 0;
+                return true;
+            } else {
+                logger.debug("Authentication failure {}", response.getContentAsString());
+                throw new MyWarmupApiException("Authentication Failed");
+            }
+        } catch (MyWarmupApiException e) {
+            failCount++;
+            if (failCount > 2) {
+                logger.error("Multiple authentication failures: {}", e.getMessage());
```

Comment:
Bindings should only log to error if something severe happened, like the detection of a bug in your code. This can be warn.

Also, it's good practice to log exceptions where they are catched. This prevents multiple logging of the same exception in different locations.

Reply:
Thanks, updated